plugins/application/elasticsearch/main.go

```diff
@@ -52,15 +52,16 @@ type Elasticsearch struct {
 	configuration *lib.AppConfig
 	logger        *logging.Logger
 	client        *lib.Client
-	buffer        *concurrent.Map
+	buffer        map[string][]string
+	bufferMutex   sync.RWMutex
```

Comment:
It looks like he is wrapping the whole write process in the lock rather than just the map. There must have been writes from other go routines happening to the buffer between the check if the index already existed (L92) and writing to the buffer later in the ReceiveEvent() function. This wouldn't necessarily be a problem with the concurrent map but rather how it was being used. That's just my thoughts looking at it, maybe Martin can explain further

Reply:
Hmmm, I'm still not sure I understand how that caused crashes, but upon re-reading with your explanation in mind, I do think I've spotted where logs were being lost! Between L96 and L102 in the old code; if two routines perform this at the same time, one of their results will be overwritten. This probably accounts for the losses I was seeing at high rates.

Reply:
It was crashing here: https://github.com/infrawatch/sg-core/blob/23ade06b40b9e1e3eb138b9b9b8aa34f9f54cc46/plugins/application/elasticsearch/main.go#L96

saying that the type of data at that index in the buffer was not a []string{} value, but a nil. It could be that another goroutine overwrote the value at that index before this type assertion happened. I'm not sure exactly where that would happen, just speculating

Reply:
Well the problem was that using concurrent.Map's method the data at index are locked only in the method itself, so yes, sometimes it happened that one goroutine fetched the data via Get and before it could work with that data other gouroutine deleted the index in map (because the map was not locked any more). That being said, concurrent.Map was not much ready for concurrency and I did not realize that when I wanted to use it for that purpose.